### PR TITLE
ENH: Optimize np.power for integer type

### PR DESCRIPTION
**Pull request by eendebakpt**
In this PR we optimize `np.power(x, n)` for integer types and a scalar argument `n`. The current implementation is a generic binary loop for the arguments. In the case `n` is a scalar (stride 0) we can optimize the  loop. 

Micro benchmark:
```
import numpy as np
import timeit

x=np.arange(100_000)
n = 20_000

dt=timeit.timeit('x**2', globals=globals(), number=n)
print(f'x**2 {1e6*dt/n}')
dt=timeit.timeit('x**3', globals=globals(), number=n)
print(f'x**3 {1e6*dt/n}')
dt=timeit.timeit('np.power(x, 1)', globals=globals(), number=n)
print(f'power(x, 1) {1e6*dt/n}')
dt=timeit.timeit('np.power(x, 2)', globals=globals(), number=n)
print(f'power(x, 2) {1e6*dt/n}')
dt=timeit.timeit('np.power(x, 3)', globals=globals(), number=n)
print(f'power(x, 3) {1e6*dt/n}')
dt=timeit.timeit('np.power(x, 4)', globals=globals(), number=n)
print(f'power(x, 4) {1e6*dt/n}')
```
Main:
```
x**2 62.292819999856874
x**3 135.96379000227898
power(x, 1) 103.64716000040062
power(x, 2) 134.2153200006578
power(x, 3) 142.9221049998887
power(x, 4) 197.58593500009738
```
PR:
```
x**2 59.52026000013575
x**3 87.76193000085186
power(x, 1) 62.03027000010479
power(x, 2) 86.0 (estimated)
power(x, 3) 86.40300999977626
power(x, 4) 126.56257500057109

```
Notes

* We could add the special case `if (in2==2)` to improve performance. For `np.power(x, 2)` we have per loop: main 143.2 us, special case exponent 2 55.7 us, no special case about 86 us.
* Related: #9363, #6399, #22651, #7786

### CODE REVIEW

**Comment by eendebakpt** [comment]
@ngoldbaum This PR is similar to #26055. Since we are dealing with integers there are no (potential) issues with accuracy and we can optimize for all exponents. On the other hand: the gain is smaller and I am not so sure about the use case for integers power.

**Comment by ngoldbaum** [comment]
It would be nice to see a proper asv benchmark for this.

File: numpy/_core/src/umath/loops.c.src [hunk 1]
```diff
@@ -474,6 +474,38 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 int
...
+                in1 *= in1;
+                if (var & 1) {
+                    out *= in1;
+                }
+                var >>= 1;
+            }
+            *((@type@ *) op1) = out;
+        }
+        return;
+    }
```
Comment (seberg) [inline numpy/_core/src/umath/loops.c.src:508]: I do wonder if we go here, if we can't create a small helper or so and not duplicate the code at all (i.e. this is just the identical code, I think.  The compiler should be able to do the same optimizations?)

Comment (eendebakpt) [reply numpy/_core/src/umath/loops.c.src:508]: It is indeed almost identical to the code on lines 532 to 539 (the only difference is the name of `var` and `x2`). We could create an method for this piece of code, probably the compiler will inline. If we want to be sure we have to create a macro. Is that what you mean?

**Comment by eendebakpt** [comment]
> It would be nice to see a proper asv benchmark for this.

I will try to do this (might take some time though, the spin command is not functioning well on my system)

**Comment by ngoldbaum** [comment]
Hop on slack at some point if you want some low-latency debugging help with that.

File: numpy/_core/src/umath/loops.c.src (hunk 1 above)
Comment (seberg) [reply numpy/_core/src/umath/loops.c.src:508]: Yeah we had a brief look yesterday and there was some confusion about added complexity so not repeating the code seemed like a good idea.  I would expect a static inline function should be sufficient in practice.

**Comment by eendebakpt** [comment]
> It would be nice to see a proper asv benchmark for this.

I added a few benchmarks for the integer `np.power` case. Results on main:
```
· Running 3 total benchmarks (1 commits * 1 environments * 3 benchmarks)
[ 0.00%] ·· Benchmarking existing-pyC__projects_misc_numpy_.venv_Scripts_python.exe
[16.67%] ··· Running (bench_ufunc.BinaryBenchInteger.time_pow--)...
[66.67%] ··· bench_ufunc.BinaryBenchInteger.time_pow                                                                                                                              ok
[66.67%] ··· ============= =============
                 dtype
             ------------- -------------
              numpy.int32    5.16±0.3ms
              numpy.int64   5.52±0.05ms
             ============= =============

[83.33%] ··· bench_ufunc.BinaryBenchInteger.time_pow_five                                                                                                                         ok
[83.33%] ··· ============= =============
                 dtype
             ------------- -------------
              numpy.int32   2.32±0.05ms
              numpy.int64    3.05±0.2ms
             ============= =============

[100.00%] ··· bench_ufunc.BinaryBenchInteger.time_pow_two                                                                                                                          ok
[100.00%] ··· ============= =============
                  dtype
              ------------- -------------
               numpy.int32    2.02±0.2ms
               numpy.int64   2.76±0.07ms
              ============= =============

```
Results for this PR:
```
· Running 3 total benchmarks (1 commits * 1 environments * 3 benchmarks)
[ 0.00%] ·· Benchmarking existing-pyC__projects_misc_numpy_.venv_Scripts_python.exe
[16.67%] ··· Running (bench_ufunc.BinaryBenchInteger.time_pow--)...
[66.67%] ··· bench_ufunc.BinaryBenchInteger.time_pow                                                                ok
[66.67%] ··· ============= ============
                 dtype
             ------------- ------------
              numpy.int32   5.07±0.3ms
              numpy.int64   5.89±0.3ms
             ============= ============

[83.33%] ··· bench_ufunc.BinaryBenchInteger.time_pow_five                                                           ok
[83.33%] ··· ============= =============
                 dtype
             ------------- -------------
              numpy.int32   1.54±0.04ms
              numpy.int64   2.24±0.09ms
             ============= =============

[100.00%] ··· bench_ufunc.BinaryBenchInteger.time_pow_two                                                            ok
[100.00%] ··· ============= =============
                  dtype
              ------------- -------------
               numpy.int32   1.17±0.06ms
               numpy.int64   1.97±0.08ms
              ============= =============
```

**Review by ngoldbaum — APPROVED** [review]
@seberg could you give this another quick look?

File: numpy/_core/src/umath/loops.c.src [hunk 2]
```diff
@@ -471,13 +471,48 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 int
 }
 /**end repeat1**/
 
+static inline @type@ _@TYPE@_squared_exponentiation_helper(@type@ base, @type@ exponent_two, int first_bit) {
```
Comment (seberg) [inline numpy/_core/src/umath/loops.c.src:474]:
```suggestion
static inline @type@
_@TYPE@_squared_exponentiation_helper(@type@ base, @type@ exponent_two, int first_bit) {
```

File: numpy/_core/src/umath/loops.c.src [hunk 3]
```diff
@@ -471,13 +471,48 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 int
...
+    if (steps[1]==0) {
+        // stride for second argument is 0
+        BINARY_DEFS_ZERO_STRIDE
+        const @type@ in2 = *(@type@ *)ip2;
+        #if @SIGNED@
+                if (in2 < 0) {
+                    npy_gil_error(PyExc_ValueError,
+                                  "Integers to negative integer powers are not allowed.");
+                    return;
+                }
```
Comment (seberg) [inline numpy/_core/src/umath/loops.c.src:500]:
```suggestion
            if (in2 < 0) {
                npy_gil_error(PyExc_ValueError,
                              "Integers to negative integer powers are not allowed.");
                return;
            }
```
Dont' mind indenting it, but seems over-indent.  (just a nitpick of course)

File: numpy/_core/src/umath/fast_loop_macros.h [hunk 4]
```diff
@@ -90,6 +90,16 @@ abs_ptrdiff(char *a, char *b)
...
 
+/** (ip1, ip2) -> (op1), for case ip2 has zero stride*/
+#define BINARY_DEFS_ZERO_STRIDE                          \
+    char *ip1 = args[0], *ip2 = args[1], *op1 = args[2]; \
+    npy_intp is1 = steps[0], os1 = steps[2];             \
+    npy_intp n = dimensions[0];                          \
+    npy_intp i;
+
+#define BINARY_LOOP_SLIDING_ZERO_STRIDE \
+    for (i = 0; i < n; i++, ip1 += is1, op1 += os1)
```
Comment (seberg) [inline numpy/_core/src/umath/fast_loop_macros.h:101]: Could you try to delete these two macros again?  The compiler should be smart enough to replace `steps[1]` with `0` in a block that says `if (steps[1] == 0)`.

File: numpy/_core/src/umath/loops.c.src [hunk 5]
```diff
@@ -471,13 +471,48 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 int
...
+        @type@ in2start = in2 >> 1;
+
+        BINARY_LOOP_SLIDING_ZERO_STRIDE {
+            @type@ in1 = *(@type@ *)ip1;
+
+            *((@type@ *) op1) = _@TYPE@_squared_exponentiation_helper(in1, in2start, first_bit);
+        }
+        return;
+    }
     BINARY_LOOP {
```
Comment (seberg) [inline numpy/_core/src/umath/loops.c.src:513]: I am happy to keep it this way, but I was half hoping that we can put this whole `BINARY_LOOP {` into a `static inline` helper and then just write:
```
if (steps[1] == 0) {
    @TYPE@_power_impl(args, dimensions, steps);
else {
    @TYPE@_power_impl(args, dimensions, steps);
}
```
Unfortunately, I suspect that the compiler cannot lift the `npy_gil_error` out of the loop so not sure it is worthwhile.
(The function could have `if (scalar_pow)` which is passed in and the compiler should optimize it away, though.)

**Review by seberg — COMMENTED** [review]
Thanks, overall I am happy with it and I think it would be fine to go ahead with the nitpicky things.

But, I am not sure I like the new macro, because I think it should be unnecessary?  And I am not sure I want to encourage the pattern.

Instead, I would love if you can see if the branching pattern works nicely here, because it is a pattern that I would prefer.

File: numpy/_core/src/umath/fast_loop_macros.h (hunk 4 above)
Comment (eendebakpt) [reply numpy/_core/src/umath/fast_loop_macros.h]: Yes, seems to work! (off course there might be some (older?) compilers that cannot do this optimization)

**Comment by ngoldbaum** [comment]
On the latest version of this PR, on my laptop:

```
| Change   | Before [dd33199e] <main>   | After [561348b4] <faster_integer_power>   |   Ratio | Benchmark (Parameter)                                               |
|----------|----------------------------|-------------------------------------------|---------|---------------------------------------------------------------------|
| -        | 1.77±0.02ms                | 1.19±0.02ms                               |    0.68 | bench_ufunc.BinaryBenchInteger.time_pow_five(<class 'numpy.int64'>) |
| -        | 1.50±0.06ms                | 964±40μs                                  |    0.64 | bench_ufunc.BinaryBenchInteger.time_pow_two(<class 'numpy.int64'>)  |
| -        | 1.29±0.03ms                | 747±10μs                                  |    0.58 | bench_ufunc.BinaryBenchInteger.time_pow_two(<class 'numpy.int32'>)  |
| -        | 2.36±0.03ms                | 995±20μs                                  |    0.42 | bench_ufunc.BinaryBenchInteger.time_pow_five(<class 'numpy.int32'>) |
```

And no slowdowns. @seberg do the low-level code changes in the final version of this look good to you?

**Comment by seberg** [comment]
Yes, thanks @eendebakpt and Nathan.

**Pull request merged**